[docs] Add `overriding-component-structure` doc to Material UI (#45186)
mui/material-ui: docs/data/material/customization/overriding-component-structure/overriding-component-structure.md

# Overriding component structure

<p class="description">Learn how to override the default DOM structure of Material UI components.</p>

Material UI components are designed to suit the widest possible range of use cases, but you may occasionally need to change how a component's structure is rendered in the DOM.

To understand how to do this, it helps to know a bit about how the API design has evolved over time, and to have an accurate mental model of the components themselves.

## Context

Prior to Material UI v6, it was not possible to override the structure of most components in the library.
Some components had `*Props` props that allowed you to pass props to a specific slot, but this pattern was not applied consistently.

In v6, those props were deprecated in favor of the `slots` and `slotProps` props, which allow for more granular control over the structure of a component and make the API more consistent across the library.

> 
> To see the available slots for a component, refer to the slots sections of the respective component API documentation.

Reviewer: Let's cut this line here and move it down to the next section.

```suggestion
```

## The mental model

A component's structure is determined by the elements that fill that component's **slots**.
Slots are most commonly filled by HTML tags, but may also be filled by React components.

> All components contain a root slot that defines their primary node in the DOM tree; more complex components also contain additional interior slots named after the elements they represent.

Reviewer: ```suggestion
All components contain a root slot that defines their primary node in the DOM tree; more complex components also contain additional interior slots named after the elements they represent.
To see the available slots for a component, refer to the slots sections of the respective component API documentation.
```

Reviewer: could also consider putting this in a callout - I don't have a strong opinion either way.

```suggestion
All components contain a root slot that defines their primary node in the DOM tree; more complex components also contain additional interior slots named after the elements they represent.

:::info
To see the available slots for a component, refer to the slots sections of the respective component API documentation.
:::
```

# Overriding component structure

<p class="description">Learn how to override the default DOM structure of Material UI components.</p>

Material UI components are designed to suit the widest possible range of use cases, but you may occasionally need to change how a component's structure is rendered in the DOM.

> To understand how to do this, it helps to have an accurate mental model of the components.

Reviewer: If you take my suggestion for rearranging the "Context" section, then I would update this line to lead in to the next section more smoothly:

```suggestion
To understand how to do this, it helps to know a bit about how the API design has evolved over time, and to have an accurate mental model of the components themselves.
```

<Autocomplete slotProps={{ root: { id: 'badge1' } }}>
```

:::warning
If both `slotProps.root` and additional props have the same keys but different values, the `slotProps.root` props will take precedence.
This does not apply to classes or the `style` prop—they will be merged instead.
:::

## Best practices

Use `component` or `slotProps.{slot}.component` prop to override the element by preserving the styles of the slot.

Use `slots` prop to replace the slot's styles and functionality with your custom component.

Overriding with `component` lets you apply the attributes of that element directly to the root.
For instance, if you override the Button's root with an `<li>` tag, you can add the `<li>` attribute `value` directly to the component.
If you did the same with `slots.root`, you would need to place this attribute on the `slotProps.root` object in order to avoid a TypeScript error.

Be mindful of your rendered DOM structure when overriding the slots of more complex components.
You can easily break the rules of semantic and accessible HTML if you deviate too far from the default structure—for instance, by unintentionally nesting block-level elements inside of inline elements.

> ## History of the pattern

Reviewer: I've added suggestions to improve this section line by line, but I think most of it can be cut because it's not really necessary for documentation purposes. I would change this header to `## Context`, move this section to the top of the doc, after the intro but before `## The mental model`, and cut the list of benefits (this would make more sense in a blog post on the topic).

Author: Sounds good to me. Updated.

Prior to Material UI v6, majority of the components lacked the ability to override their structure.
Some components had `*Props` props that allowed you to pass props to a specific slot, but this was not a consistent pattern across the library.

In v6, those props were deprecated in favor of the `slots` and `slotProps` props, which allow for more granular control over the structure of a component and make the API more consistent across the library.

> The summary of the benefits of the new pattern is as follows:

Reviewer: ```suggestion
The benefits of the new pattern include:
```

> - The `slotProps` provided to [theme default props](/material-ui/customization/theme-components/#theme-default-props) are merged with the instance `slotProps` prop. This is useful for building custom themes or use Material UI as a design system component library.

Reviewer: I'm not sure if I understand the first sentence—does it make sense as "instance's"?

```suggestion
- `slotProps` provided to [the default theme props](/material-ui/customization/theme-components/#theme-default-props) are merged with the instance's `slotProps` prop. This is useful for building custom themes or using Material UI as the basis for a design system component library.
```

> - Most components have a `slots` prop that allows developers to replace the entire slot with a custom one. For example, developers can remove the popper behavior from the Autocomplete component by replacing `popper` slot with a `div` as [shown above](#the-slots-prop).

Reviewer: ```suggestion
- Most components have a `slots` prop that allows you to replace the entire slot with a custom element. For example, you can remove the popper behavior from the Autocomplete component by replacing the `popper` slot with a `<div>` as [shown above](#the-slots-prop).
```

> - Integrate better with [Tailwind CSS](https://tailwindcss.com/) because developers can provide utility classes to the slots through `slotProps.*.className`.

Reviewer: ```suggestion
- Offers better integration with [Tailwind CSS](https://tailwindcss.com/) because developers can provide utility classes to the slots through `slotProps.*.className`.
```

> To see the available slots for a component, refer to the component API documentation at slots section.

Reviewer: ```suggestion
To see the available slots for a component, refer to the slots sections of the respective component API documentation.
```

<Autocomplete slotProps={{ root: { id: 'badge1' } }}>
```

:::warning
If both `slotProps.root` and additional props have the same keys but different values, the `slotProps.root` props will take precedence.
This does not apply to classes or the `style` prop—they will be merged instead.
:::

## Best practices

Use `component` or `slotProps.{slot}.component` prop to override the element by preserving the styles of the slot.

Use `slots` prop to replace the slot's styles and functionality with your custom component.

Overriding with `component` lets you apply the attributes of that element directly to the root.
For instance, if you override the Button's root with an `<li>` tag, you can add the `<li>` attribute `value` directly to the component.
If you did the same with `slots.root`, you would need to place this attribute on the `slotProps.root` object in order to avoid a TypeScript error.

Be mindful of your rendered DOM structure when overriding the slots of more complex components.
You can easily break the rules of semantic and accessible HTML if you deviate too far from the default structure—for instance, by unintentionally nesting block-level elements inside of inline elements.

> Material UI components automatically correct semantically incorrect HTML—see [Automatic adjustment](/material-ui/main-features/automatic-adjustment/) for details.

Author: ```suggestion
```
